File: java/jython/JavaWrapperOverhead.java
import org.python.util.PythonInterpreter;

import java.util.Properties;

public class JavaWrapperOverhead {

  public static class Calc {
    public int square(int n) {
      return n * n;
    }
  }
  
  public static void main(String[] args) {
    Properties properties = new Properties();
    properties.setProperty("python.cachedir.skip", "true");
    properties.setProperty("python.home", "jython-src");

    PythonInterpreter.initialize(System.getProperties(),
                                 properties,
                                 new String[]{});
    PythonInterpreter inter = new PythonInterpreter();
    inter.exec("import java_wrapper_overhead");
    long start, end;
    start = System.currentTimeMillis();
    inter.exec("java_wrapper_overhead.UseInherits()");
    end = System.currentTimeMillis();
    System.out.println("time: " + (end - start) + "[ms]");
    start = System.currentTimeMillis();
    inter.exec("java_wrapper_overhead.NoInherits()");
    end = System.currentTimeMillis();
    System.out.println("time: " + (end - start) + "[ms]");
    start = System.currentTimeMillis();
    inter.exec("java_wrapper_overhead.Proxy()");
    end = System.currentTimeMillis();
    System.out.println("time: " + (end - start) + "[ms]");
  }
}

File: java/jython/python/java_wrapper_overhead.py
import JavaWrapperOverhead.Calc

LOOP = 1000 * 1000

# time: 5947[ms]
# time: 723[ms]
# time: 1684[ms]

def UseInherits():
    class PyCalc(JavaWrapperOverhead.Calc):
        pass
    for i in xrange(LOOP):
        PyCalc().square(i)

def NoInherits():
    for i in xrange(LOOP):
        JavaWrapperOverhead.Calc().square(i)

def Proxy():
    class PyCalc(object):
        def __init__(self):
            self.__calc = JavaWrapperOverhead.Calc()
        def square(self, n):
            return self.__calc.square(n)
            
    for i in xrange(LOOP):
        PyCalc().square(i)
